# 펭귄 몸무게 예측하기 실습

## 1.  데이터 불러오기

사용할 라이브러리들을 import하고, 학습시킬 데이터를 불러와서 train 변수에 저장

In [168]:
import numpy as np
import pandas as pd

# 데이터 불러오기
train = pd.read_csv('./train_penguins.csv',index_col=0)

## 2. 데이터 살펴보기

head()와 dtypes를 이용해 데이터를 살펴봅시당.

In [169]:
train.head()

Species  Island  ... Delta 13 C (o/oo)  Body Mass (g)
0          Gentoo penguin (Pygoscelis papua)  Biscoe  ...         -25.42826         5100.0
1          Gentoo penguin (Pygoscelis papua)  Biscoe  ...         -26.61788         4200.0
2          Gentoo penguin (Pygoscelis papua)  Biscoe  ...         -25.38017         5150.0
3          Gentoo penguin (Pygoscelis papua)  Biscoe  ...         -25.76147         5500.0
4  Chinstrap penguin (Pygoscelis antarctica)   Dream  ...         -24.68741         4100.0

[5 rows x 10 columns]

In [170]:
train.dtypes

Species                 object
Island                  object
Clutch Completion       object
Culmen Length (mm)     float64
Culmen Depth (mm)      float64
Flipper Length (mm)    float64
Sex                     object
Delta 15 N (o/oo)      float64
Delta 13 C (o/oo)      float64
Body Mass (g)          float64
dtype: object

 <br><br>
 **펭귄 데이터의 컬럼 정보**

<br>

- **Species** : 펭귄의 종 / 카테고리형 데이터 

- **Island** : 시료를 채취한 섬 / 카테고리형 데이터

- **Clutch Completion** : 연구 둥지가 <u>완전한 클러치(2개의 알)</u>로 관찰되었는지 여부 / 카테고리형 데이터

- **Culmen Length (mm)** : 부리의 등쪽 능선의 길이 (mm) / 수치형 데이터

- **Culmen Depth (mm)** : 부리의 등쪽 능선의 깊이 (mm) / 수치형 데이터

- **Flipper Length (mm)** : 펭귄 지느러미의 길이 (mm) / 수치형 데이터

- **Sex** : 펭귄의 성별 / 카테고리형 데이터

- **Delta 15 N (o/oo)** : 안정 동위원소 15N:14N의 비율 / 수치형 데이터

- **Delta 13 C (o/oo)** : 안정 동위원소 13C:12C의 비율 / 수치형 데이터

- **Body Mass (g)** : 펭귄의 체질량 (g) / 수치형 데이터

<br><br>
 출처) https://allisonhorst.github.io/palmerpenguins/reference/penguins_raw.html

<br>

## 3.  결측치 확인

info()를 이용해 결측치가 존재하는지 확인해보겠습니다.

In [171]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 240 entries, 0 to 239
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Species              240 non-null    object 
 1   Island               240 non-null    object 
 2   Clutch Completion    240 non-null    object 
 3   Culmen Length (mm)   239 non-null    float64
 4   Culmen Depth (mm)    239 non-null    float64
 5   Flipper Length (mm)  239 non-null    float64
 6   Sex                  231 non-null    object 
 7   Delta 15 N (o/oo)    230 non-null    float64
 8   Delta 13 C (o/oo)    231 non-null    float64
 9   Body Mass (g)        239 non-null    float64
dtypes: float64(6), object(4)
memory usage: 20.6+ KB


In [172]:
train.isna().sum()

Species                 0
Island                  0
Clutch Completion       0
Culmen Length (mm)      1
Culmen Depth (mm)       1
Flipper Length (mm)     1
Sex                     9
Delta 15 N (o/oo)      10
Delta 13 C (o/oo)       9
Body Mass (g)           1
dtype: int64

Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm), Sex, Delta 15 N (o/oo), Delta 13 C (o/oo) Body Mass (g) 컬럼에 결측치가 존재합니다.

<br>

## 3-1  결측치 처리

In [173]:
train[train.isna().sum(axis=1) > 0]

Species     Island  ... Delta 13 C (o/oo)  Body Mass (g)
36           Gentoo penguin (Pygoscelis papua)     Biscoe  ...         -26.23886         4725.0
42         Adelie Penguin (Pygoscelis adeliae)  Torgersen  ...               NaN         3700.0
58         Adelie Penguin (Pygoscelis adeliae)      Dream  ...               NaN         3900.0
67         Adelie Penguin (Pygoscelis adeliae)      Dream  ...               NaN         4650.0
77         Adelie Penguin (Pygoscelis adeliae)  Torgersen  ...               NaN         3700.0
94           Gentoo penguin (Pygoscelis papua)     Biscoe  ...         -26.81540         4650.0
98   Chinstrap penguin (Pygoscelis antarctica)      Dream  ...         -23.78767         3950.0
100        Adelie Penguin (Pygoscelis adeliae)  Torgersen  ...               NaN         3475.0
121        Adelie Penguin (Pygoscelis adeliae)  Torgersen  ...         -25.09368         4250.0
125          Gentoo penguin (Pygoscelis papua)     Biscoe  ...         -25.69327         4100.0
126          Gentoo penguin (Pygoscelis papua)     Biscoe  ...         -26.18444         4875.0
196        Adelie Penguin (Pygoscelis adeliae)  Torgersen  ...               NaN         3750.0
198        Adelie Penguin (Pygoscelis adeliae)  Torgersen  ...               NaN         3800.0
216        Adelie Penguin (Pygoscelis adeliae)  Torgersen  ...         -25.21315         3300.0
228        Adelie Penguin (Pygoscelis adeliae)  Torgersen  ...               NaN            NaN
233        Adelie Penguin (Pygoscelis adeliae)      Dream  ...               NaN         3425.0

[16 rows x 10 columns]

결측치가 존재하는 행을 출력했을 때, Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm), Body Mass (g), Sex, Delta 15 N (o/oo), Delta 13 C (o/oo)	컬럼에서 NaN 값을 확인할 수 있습니다.

<br>

전체적으로 데이터가 없는 행은 삭제를 하고,
카테고리형 데이터에 결측치가 존재하는 경우는 행을 삭제하고, 

수치형 데이터에 결측치가 존재하는 경우는 0값을 채워주겠습니다.

In [156]:
train_preprocessed = train.dropna(subset=['Culmen Length (mm)'])

In [157]:
# 카테고리형 데이터에서 결측치에 해당하는 행 삭제
train_preprocessed = train.dropna(subset=['Sex'])

In [158]:
# 수치형 데이터에서 결측치에 0값 넣기
train_preprocessed = train_preprocessed.fillna(0)

In [159]:
# 확인
train_preprocessed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 231 entries, 0 to 239
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Species              231 non-null    object 
 1   Island               231 non-null    object 
 2   Clutch Completion    231 non-null    object 
 3   Culmen Length (mm)   231 non-null    float64
 4   Culmen Depth (mm)    231 non-null    float64
 5   Flipper Length (mm)  231 non-null    float64
 6   Sex                  231 non-null    object 
 7   Delta 15 N (o/oo)    231 non-null    float64
 8   Delta 13 C (o/oo)    231 non-null    float64
 9   Body Mass (g)        231 non-null    float64
dtypes: float64(6), object(4)
memory usage: 19.9+ KB


결측치가 존재하지 않습니다.

<br>

## 4. 레이블 인코딩

아직 범주형 변수들이 문자열의 형태로 존재하고 있기 때문에 모델 학습이 가능하도록 문자열 형태의 범주형 변수들을 숫자 형태로 인코딩해주어야 합니다.

1. 방법 1
- 레이블 인코딩을 하기 위한 dictionary map 생성 함수
- 각 범주형 변수에 인코딩 값을 부여하는 함수 만들어서 transform

In [146]:
# 레이블 인코딩을 하기 위한 dictionary map 생성 함수
def make_label_map(dataframe):
    label_maps = {}
    for col in dataframe.columns:
        if dataframe[col].dtype=='object': 
            label_map = {'unknown':0} 
            for i, key in enumerate(train[col].unique()):
                label_map[key] = i+1
            label_maps[col] = label_map
    return label_maps

# 각 범주형 변수에 인코딩 값을 부여하는 함수
def label_encoder(dataframe, label_map):
    for col in dataframe.columns:
        if dataframe[col].dtype=='object':
            dataframe[col] = dataframe[col].map(label_map[col])
            dataframe[col] = dataframe[col].fillna(label_map[col]['unknown']) #혹시 모를 결측값은 unknown의 값으로 채워줍니다.
    return dataframe

- **enumerate(train[col].unique())**

  enumerate()는 인덱스와 원소로 이루어진 튜플을 생성하는 함수입니다. <br>unique는 고유값을 찾아주는 함수입니다.

- **dataframe[col].map(label_map[col])**

  map 메서드를 이용하면 기존값(key)을 대체값(value)으로 변경할 수 있습니다.

In [147]:
# 전처리 완료한 train 데이터 : train_preprocessed
label_map = make_label_map(train_preprocessed) # train_preprocessed 사용해 label map 생성
labeled_train = label_encoder(train_preprocessed, label_map) # train_preprocessed 라벨 인코딩

2. 방법 2
- LabelEncoder 사용

In [148]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 240 entries, 0 to 239
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Species              240 non-null    object 
 1   Island               240 non-null    object 
 2   Clutch Completion    240 non-null    object 
 3   Culmen Length (mm)   239 non-null    float64
 4   Culmen Depth (mm)    239 non-null    float64
 5   Flipper Length (mm)  239 non-null    float64
 6   Sex                  231 non-null    object 
 7   Delta 15 N (o/oo)    230 non-null    float64
 8   Delta 13 C (o/oo)    231 non-null    float64
 9   Body Mass (g)        239 non-null    float64
dtypes: float64(6), object(4)
memory usage: 20.6+ KB


In [160]:
from sklearn.preprocessing import LabelEncoder

train2 =train.copy()

le = LabelEncoder()
le.fit(train2['Species'])
train2['Species'] = le.transform(train2['Species'])

le2 = LabelEncoder()
le2.fit(train2['Island'])
train2['Island'] = le2.transform(train2['Island'])

le3= LabelEncoder()
le3.fit(train2['Clutch Completion'])
train2['Clutch Completion'] = le3.transform(train2['Clutch Completion'])

le4 = LabelEncoder()
le4.fit(train2['Sex'])
train2['Sex'] = le4.transform(train2['Sex'])

In [163]:
from collections import defaultdict

class MultiCoLableEncoder:
    def __init__(self):
        self.encoder_dict = defaultdict(LabelEncoder)
    def fit_transform(self, X:pd.DataFrame,columns:list):
        if not isinstance(columns,list):
            columns = [columns]
        output = X.copy()
        output[columns] = X[columns].apply(lambda x:self.encoder_dict[x.name].fit_transform(x))
        return output

In [165]:
mcle = MultiCoLableEncoder()
transformed_data = mcle.fit_transform(train,columns=['Species','Island','Clutch Completion','Sex'])

## 5.  모델링

모델링을 하기에 앞서, 회귀 평가 지표로 사용할 RMSE를 출력해주는 함수로 만들어주겠습니다.

사이킷런은 RMSE를 제공하지 않기 때문에 RMSE를 구하기 위해서는 MSE에 제곱근을 씌워서 계산하는 함수를 직접 만들어야 합니다.

In [96]:
import numpy as np

def RMSE(true, pred):
    score = np.sqrt(np.mean(np.square(true-pred)))
    return score

<br>

K-Fold 교차검증 방식으로 모델링을 해보도록 하겠습니다.


<br>

target 변수에 종속변수인 'Body Mass (g)'를 저장하고,

feature 변수에 불필요한 종속변수 'Body Mass (g)'를 제거한 데이터를 저장해줍니다.

In [97]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import KFold

target = labeled_train['Body Mass (g)']
print(target.shape)

# 학습에 불필요한 변수 제거
feature = labeled_train.drop(['Body Mass (g)'], axis=1)
print(feature.shape)

(231,)
(231, 9)


<br>

LinearRegression와 5개의 폴드 세트로 분리하는 KFold 객체를 생성해줍니다.

In [98]:
lr = LinearRegression()

# 폴드 개수 할당
kfold = KFold(n_splits=5)

- **LinearRegression()**

  LinearRegression 클래스를 이용하면 선형 회귀 모델을 만들 수 있습니다.

- **KFold(n_splits=5)**

  KFold 함수를 이용해 폴드 개수를 할당해줄 수 있습니다. 

  일반적으로 5, 10 등의 값을 사용합니다.

<br>

각 폴드마다 다른 x_train과 y_train을 이용해 모델을 학습하고, x_test를 예측해 예측값을 y_test와 비교하여 RMSE를 구해보도록 하겠습니다.

In [99]:
cv_rmse = []
n_iter = 0

for train_index, test_index, in kfold.split(feature):

    x_train, x_test = feature.iloc[train_index], feature.iloc[test_index]
    y_train, y_test = target.iloc[train_index], target.iloc[test_index]

    # 모델 학습
    lr = lr.fit(x_train, y_train)

    # 테스트셋 예측
    pred = lr.predict(x_test)

    # 반복 횟수 1회 증가
    n_iter += 1

    error = RMSE(y_test, pred)
    train_size = x_train.shape[0]
    test_size = x_test.shape[0]

    print('\n{0}번째 교차 검증 RMSE : {1}, 학습 데이터 크기 : {2}, 검증 데이터 크기 : {3}'.format(n_iter, error, train_size, test_size))
    print('{0}번째 검증 세트 인덱스 : {1}'.format(n_iter, test_index))
    cv_rmse.append(error)

print('\n==> 이 방정식의 평균 RMSE는 {}입니다.'.format(np.mean(cv_rmse)))


1번째 교차 검증 RMSE : 330.6091274698889, 학습 데이터 크기 : 184, 검증 데이터 크기 : 47
1번째 검증 세트 인덱스 : [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46]

2번째 교차 검증 RMSE : 359.52128166381306, 학습 데이터 크기 : 185, 검증 데이터 크기 : 46
2번째 검증 세트 인덱스 : [47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70
 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92]

3번째 교차 검증 RMSE : 496.9885201420895, 학습 데이터 크기 : 185, 검증 데이터 크기 : 46
3번째 검증 세트 인덱스 : [ 93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108 109 110
 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128
 129 130 131 132 133 134 135 136 137 138]

4번째 교차 검증 RMSE : 343.577352026888, 학습 데이터 크기 : 185, 검증 데이터 크기 : 46
4번째 검증 세트 인덱스 : [139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156
 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174
 175 176 177 178 179 180 181 182 183 184]

- **kfold.split(feature)**

  split() 함수를 호출해 전체 펭귄 데이터를 5개의 폴드 데이터 세트로 분리합니다.

- **lr.fit(x_train, y_train)**

  LinearRegression 클래스의 fit 메소드를 사용해 훈련시킵니다.

- **RMSE(y_test, pred)**

  예측값(pred)과 실제값(y_test)의 차이를 기반으로 하는 RMSE를 구합니다.



<br>

교차 검증을 더 편리하게 수행할 수 있게 해주는 API인 cross_val_score()를 이용해 교차 검증으로 MSE와 RMSE를 추정해보겠습니다.

In [100]:
# cross_val_score()로 구현하기
from sklearn.model_selection import cross_val_score

neg_mse_scores = cross_val_score(lr, feature, target, scoring="neg_mean_squared_error", cv=5)
rmse_scores = np.sqrt(-1 * neg_mse_scores)
avg_rmse = np.mean(rmse_scores)

print(' 5 folds 의 개별 Negative MSE scores : ', np.round(neg_mse_scores))
print(' 5 folds 의 개별 RMSE scores : ', np.round(rmse_scores))
print(' 5 folds 의 평균 RMSE : ', np.round(avg_rmse, 3))

 5 folds 의 개별 Negative MSE scores :  [-109302. -129256. -246998. -118045.  -96823.]
 5 folds 의 개별 RMSE scores :  [331. 360. 497. 344. 311.]
 5 folds 의 평균 RMSE :  368.372


- **cross_val_score(lr, feature, target, scoring="neg_mean_squared_error", cv=5)**

  scoring은 예측 성능 평가 지표입니다. "neg_mean_squared_error"는 MSE에 음수를 취한 값입니다.

  cv는 교차 검증 폴드 수를 결정합니다.

- **np.sqrt(-1 * neg_mse_scores)**

  RMSE를 구하기 위해서 음수 값인 neg_mse_scores에 -1을 곱해주고 제곱근을 취해줍니다.

<br>

## 6. 예측하기

예측을 진행 할 test.csv 데이터를 불러옵니다.

In [101]:
# 데이터 불러오기
test = pd.read_csv('./test_penguins.csv',index_col=0)

# 데이터 확인
test.head()

Species  Island  ... Delta 13 C (o/oo)  Body Mass (g)
0        Adelie Penguin (Pygoscelis adeliae)   Dream  ...         -25.06020         3300.0
1          Gentoo penguin (Pygoscelis papua)  Biscoe  ...         -26.60436         5550.0
2  Chinstrap penguin (Pygoscelis antarctica)   Dream  ...         -24.38933         4050.0
3          Gentoo penguin (Pygoscelis papua)  Biscoe  ...         -26.70968         4950.0
4          Gentoo penguin (Pygoscelis papua)  Biscoe  ...         -26.67799         5300.0

[5 rows x 10 columns]

In [109]:
test.drop('Body Mass (g)',axis=1,inplace=True)

<br>

test.csv에도 train.csv와 동일한 전처리 과정을 진행해줍니다.

In [110]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 104 entries, 0 to 103
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Species              104 non-null    object 
 1   Island               104 non-null    object 
 2   Clutch Completion    104 non-null    object 
 3   Culmen Length (mm)   103 non-null    float64
 4   Culmen Depth (mm)    103 non-null    float64
 5   Flipper Length (mm)  103 non-null    float64
 6   Sex                  104 non-null    object 
 7   Delta 15 N (o/oo)    100 non-null    float64
 8   Delta 13 C (o/oo)    100 non-null    float64
dtypes: float64(5), object(4)
memory usage: 8.1+ KB


Sex, Delta 15 N (o/oo), Delta 13 C (o/oo) 컬럼에 결측값이 존재합니다.

<br>

train.csv는 Sex 컬럼의 결측치가 존재하는 행을 삭제했지만 test.csv의 행은 삭제하면 안되므로, 

train.csv의 분포에서 개수가 더 큰 "MALE" 값을 임의로 부여합니다.

In [111]:
train[['Sex']].value_counts()

Sex   
FEMALE    116
MALE      115
dtype: int64

In [112]:
test['Sex'] = test['Sex'].fillna("MALE")

<br>

train.csv와 마찬가지로 수치형 변수의 결측값에는 0값을 채워줍니다.

In [113]:
test_preprocessed = test.fillna(0)
test_preprocessed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 104 entries, 0 to 103
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Species              104 non-null    object 
 1   Island               104 non-null    object 
 2   Clutch Completion    104 non-null    object 
 3   Culmen Length (mm)   104 non-null    float64
 4   Culmen Depth (mm)    104 non-null    float64
 5   Flipper Length (mm)  104 non-null    float64
 6   Sex                  104 non-null    object 
 7   Delta 15 N (o/oo)    104 non-null    float64
 8   Delta 13 C (o/oo)    104 non-null    float64
dtypes: float64(5), object(4)
memory usage: 8.1+ KB


결측치가 존재하지 않습니다.

<br>

test.csv에도 레이블 인코딩을 동일하게 진행해줍니다.

In [114]:
labeled_test = label_encoder(test_preprocessed, label_map)
labeled_test.head()

Species  Island  Clutch Completion  ...  Sex  Delta 15 N (o/oo)  Delta 13 C (o/oo)
0        3       2                  1  ...    2            9.66523          -25.06020
1        1       1                  1  ...    1            8.46894          -26.60436
2        2       2                  1  ...    1            9.02330          -24.38933
3        1       1                  1  ...    2            7.96935          -26.70968
4        1       1                  1  ...    1            8.55868          -26.67799

[5 rows x 9 columns]

In [166]:
from sklearn.preprocessing import LabelEncoder

test['Species'] = le.transform(test['Species'])
test['Island'] = le2.transform(test['Island'])
test['Clutch Completion'] = le3.transform(test['Clutch Completion'])
test['Sex'] = le4.transform(test['Sex'])

<br>

In [115]:
labeled_test.shape

(104, 9)

<br>

전처리가 모두 완료된 labeled_test 데이터로 예측해줍니다. 

In [116]:
predict_test = lr.predict(labeled_test)
predict_test

array([3307.65229556, 5414.59631685, 4409.68439773, 4672.26539464,
       5244.13429545, 5262.89869547, 3974.54818493, 4272.02000566,
       5608.60865995, 3429.68534699, 4420.89183642, 3403.72280612,
       4218.93653598, 3714.03973465, 3123.78876453, 3878.41490553,
       4408.486941  , 4633.66787818, 4368.12878931, 4485.29622546,
       4197.60044966, 3965.11753655, 4556.53785859, 3172.17128263,
       4009.76455836, 4584.48982291, 4039.28141882, 4832.35890971,
       4208.79366503, 4235.45753801, 4658.44176384, 3633.78110103,
       3460.11668128, 4659.76053767, 4210.28651118, 3386.52081855,
       4672.11933313, 3458.28566796, 3648.30007854, 3655.81549312,
       3486.94781326, 3540.73230978, 5391.86238431,  239.53675076,
       3975.56233919, 5288.34035622, 4589.27553311, 5235.38156076,
       3247.78308913, 4345.68504571, 4077.16961698, 5465.54260281,
       5159.65345966, 3428.07957196, 5827.14437758, 5342.08731059,
       4171.55648039, 3315.48840127, 3901.76440612, 5429.28024

<br>

이렇게 선형 회귀 기법으로 펭귄의 몸무게를 예측해봤습니다.

평균 RMSE 값이 368.372로 매우 높게 나와 좋은 모델이라고 할 수는 없지만,더 좋은 모델로 수정할 수 있을 것입니다!

<br><br>

## **2. 규제 선형 모델**

<br>

### **1) 릿지 회귀 (Ridge)**

W의 제곱에 대해 페널티를 부여하는 L2 규제를 선형 회귀에 적용한 것이 **릿지(Ridge)** 회귀입니다.

사이킷런은 **Ridge** 클래스를 통해 릿지 회귀를 구현하고, Ridge 클래스의 주요 생성 파라미터 alpha는 릿지 회귀의 alpha L2 규제 계수에 해당합니다. 

alpha 값을 증가시키면 회귀 계수가 0에 가까워집니다.

<br>

펭귄 몸무게 예측 예제를 Ridge 클래스를 이용해 다시 예측해보고, 예측 성능을 cross_val_score()로 평가해보겠습니다.

In [117]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import cross_val_score

# alpha=10
ridge = Ridge(alpha=10)
neg_mse_scores = cross_val_score(ridge, feature, target, scoring="neg_mean_squared_error", cv=5)
rmse_scores = np.sqrt(-1 * neg_mse_scores)
avg_rmse = np.mean(rmse_scores)

In [118]:
print(' 5 folds 의 개별 Negative MSE scores: ', np.round(neg_mse_scores, 3))
print(' 5 folds 의 개별 RMSE scores : ', np.round(rmse_scores, 3))
print(' 5 folds 의 평균 RMSE : {0:.3f} '.format(avg_rmse))

 5 folds 의 개별 Negative MSE scores:  [-100770.35  -131053.493 -225176.603 -135050.741  -86674.265]
 5 folds 의 개별 RMSE scores :  [317.443 362.013 474.528 367.493 294.405]
 5 folds 의 평균 RMSE : 363.176 


- **Ridge(alpha=10)**

  Ridge 클래스의 alpha 값을 10으로 지정해 릿지 회귀를 구현합니다.

- **cross_val_score(ridge, feature, target, scoring="neg_mean_squared_error", cv=5)**

  scoring은 예측 성능 평가 지표입니다. "neg_mean_squared_error"는 MSE에 음수를 취한 값입니다.

  cv는 교차 검증 폴드 수를 결정합니다.

- **np.sqrt(-1 * neg_mse_scores)**

  RMSE를 구하기 위해서 음수 값인 neg_mse_scores에 -1을 곱해주고 제곱근을 취해줍니다.

<br>


alpha 값의 변화에 따른 RMSE와 릿지 회귀 계수 값의 변화를 살펴보겠습니다.

In [119]:
# alpha 값의 변화에 따른 평균 RMSE의 변화

ridge_alphas = [0, 0.1, 1, 10, 100]

print('####### Ridge #######\n')

for alpha in ridge_alphas :
    ridge = Ridge(alpha = alpha)

    neg_mse_scores = cross_val_score(ridge, feature, target, scoring="neg_mean_squared_error", cv = 5)
    avg_rmse = np.mean(np.sqrt(-1 * neg_mse_scores))
    print('alpha {0} 일 때 5 folds 의 평균 RMSE : {1:.3f} '.format(alpha, avg_rmse))

####### Ridge #######

alpha 0 일 때 5 folds 의 평균 RMSE : 368.372 
alpha 0.1 일 때 5 folds 의 평균 RMSE : 368.196 
alpha 1 일 때 5 folds 의 평균 RMSE : 366.851 
alpha 10 일 때 5 folds 의 평균 RMSE : 363.176 
alpha 100 일 때 5 folds 의 평균 RMSE : 376.154 


<br>

In [120]:
# alpha 값의 변화에 따른 회귀 계수의 변화

coeff_df = pd.DataFrame()

for pos, alpha in enumerate(ridge_alphas) :
    ridge = Ridge(alpha = alpha)
    ridge.fit(feature, target)
 
    coeff = pd.Series(data=ridge.coef_, index=feature.columns)
    colname = 'alpha:'+str(alpha)
    coeff_df[colname] = coeff

ridge_alphas = [0,  0.1, 1, 10, 100]
sort_column = 'alpha:'+str(ridge_alphas[0])
coeff_df.sort_values(by=sort_column, ascending=False)

alpha:0   alpha:0.1     alpha:1    alpha:10   alpha:100
Flipper Length (mm)   29.082844   29.169443   29.897896   34.314271   41.764970
Clutch Completion     -7.387869   -7.204686   -5.765509   -0.961186   -2.282067
Delta 13 C (o/oo)     -7.974664   -7.925173   -7.511890   -5.092580    0.916309
Culmen Length (mm)    -9.228013   -9.038925   -7.466795    1.243996   11.961309
Culmen Depth (mm)    -21.604058  -21.596776  -21.405645  -14.940522    8.924421
Delta 15 N (o/oo)    -41.620844  -41.605755  -41.484427  -40.782001  -31.017615
Island               -95.334313  -95.737998  -99.007381 -112.022998  -71.948597
Species             -224.156933 -221.835259 -202.688799 -103.511426  -13.930833
Sex                 -573.114156 -570.694116 -549.987683 -409.899150 -126.651607

- **enumerate(ridge_alphas)**

  인덱스와 원소를 동시에 리턴합니다.

- **coeff_df.sort_values(by=sort_column, ascending=False)**

  alpha:0 컬럼의 값을 기준으로 내림차순 정렬했습니다.



<br>

alpha 값이 0, 0.1, 1, 10, 100으로 변화함에 따라 평균 RMSE의 값과 회귀 계수가 어떻게 변화하는지 살펴보았습니다.

alpha가 1일 때 334.030으로 가장 좋은 평균 RMSE를 보여주며,
규제가 없는 LinearRegression의 평균 RMSE인 336.435보다 더 뛰어난 예측 성능을 보여줍니다.

alpha 값을 계속 증가시킬수록 회귀 계수 값은 0에 가까워짐을 확인할 수 있습니다.

하지만 릿지 회귀의 경우, 회귀 계수를 0으로 만들지는 않습니다.

<br>

### **2) 라쏘 회귀 (Lasso)**

W의 절댓값에 페널티를 부여하는 L1 규제를 선형 회귀에 적용한 것이 **라쏘(Lasso)** 회귀입니다.

L1 규제는 불필요한 회귀 계수를 급격하게 감소시켜 0으로 만들고 제거합니다. 이러한 점에서 L1 규제는 적절한 피처만 회귀에 포함시키는 피처 선택의 특성을 가지고 있습니다. 

사이킷런은 **Lasso** 클래스를 통해 라쏘 회귀를 구현했습니다. Lasso 클래스의 주요 생성 파라미터 alpha는 라쏘 회귀의 alpha L1 규제 계수에 해당합니다.

<br>

펭귄 몸무게 예측 예제를 Lasso 클래스를 이용해 다시 예측해보고, 
alpha 값의 변화에 따른 RMSE와 라쏘 회귀 계수 값의 변화를 살펴보겠습니다.

In [121]:
# alpha 값의 변화에 따른 평균 RMSE의 변화

from sklearn.linear_model import Lasso
from sklearn.model_selection import cross_val_score

print('####### Lasso #######\n')

lasso_alphas = [0.07, 0.1, 0.5, 1, 3]

for alpha in lasso_alphas :
    lasso = Lasso(alpha = alpha)

    neg_mse_scores = cross_val_score(lasso, feature, target, scoring="neg_mean_squared_error", cv = 5)
    avg_rmse = np.mean(np.sqrt(-1 * neg_mse_scores))
    print('alpha {0} 일 때 5 folds 의 평균 RMSE : {1:.3f} '.format(alpha, avg_rmse))

####### Lasso #######

alpha 0.07 일 때 5 folds 의 평균 RMSE : 368.263 
alpha 0.1 일 때 5 folds 의 평균 RMSE : 368.217 
alpha 0.5 일 때 5 folds 의 평균 RMSE : 367.610 
alpha 1 일 때 5 folds 의 평균 RMSE : 366.885 
alpha 3 일 때 5 folds 의 평균 RMSE : 364.480 


<br>

In [122]:
# alpha 값의 변화에 따른 회귀 계수의 변화

coeff_df = pd.DataFrame()
lasso_alphas = [0.07, 0.1, 0.5, 1, 3]

for pos, alpha in enumerate(lasso_alphas) :
    lasso = Lasso(alpha = alpha)
    lasso.fit(feature, target)
 
    coeff = pd.Series(data=lasso.coef_, index=feature.columns)
    colname = 'alpha:'+str(alpha)
    coeff_df[colname] = coeff

sort_column = 'alpha:'+str(lasso_alphas[0])
coeff_df.sort_values(by=sort_column, ascending=False)

alpha:0.07   alpha:0.1   alpha:0.5     alpha:1     alpha:3
Flipper Length (mm)   29.129739   29.149837   29.418134   29.730867   30.920415
Clutch Completion     -6.654898   -6.340767   -2.151176   -0.000000   -0.000000
Delta 13 C (o/oo)     -7.912180   -7.885402   -7.528091   -7.063697   -5.154565
Culmen Length (mm)    -9.140183   -9.102542   -8.600180   -7.973273   -5.466018
Culmen Depth (mm)    -21.691038  -21.728315  -22.226535  -22.805215  -25.003936
Delta 15 N (o/oo)    -41.496428  -41.443107  -40.731669  -39.784139  -35.827146
Island               -95.436887  -95.480847  -96.067615  -96.612479  -98.268244
Species             -222.909590 -222.375011 -215.237702 -206.777576 -174.173329
Sex                 -572.272495 -571.911783 -567.099775 -561.129000 -537.356667

<br>

alpha 값이 0.07, 0.1, 0.5, 1, 3으로 변화함에 따라 평균 RMSE의 값과 회귀 계수가 어떻게 변화하는지 살펴보았습니다.

alpha가 3일 때 332.923으로 가장 좋은 평균 RMSE를 보여주며,
앞의 릿지 회귀의 평균 RMSE 334.030보다 더 좋은 수치입니다.

현재 출력값에는 잘 나타나지 않았지만, 라쏘 회귀의 특성상 불필요한 회귀 계수를 0으로 만들기 때문에 회귀 계수가 0인 경우가 종종 나타납니다.

<br>

### **3) 엘라스틱넷 회귀 (Elastic Net)**


**엘라스틱넷(Elastic Net)** 회귀는 L2 규제와 L1 규제를 결합한 회귀입니다.

라쏘 회귀는 서로 상관관계가 높은 피처들의 경우 중요 피처만을 선택하고 다른 피처들의 회귀 계수는 모두 0으로 만드는 성향이 강합니다. 하지만 이러한 성향으로 alpha 값에 따라 회귀 계수의 값이 급격히 변동할 수도 있는데, 이를 완화하기 위해 L2 규제가 추가되었습니다.

하지만 엘라스틱넷의 규제는 L1 규제와 L2 규제가 결합된 규제이기 때문에 수행시간이 상대적으로 오래 걸린다는 단점이 있습니다.

<br>


사이킷런은 **ElasticNet** 클래스를 통해서 엘라스틱넷 회귀를 구현합니다. 

ElasticNet 클래스의 주요 생성 파라미터는 *alpha*와 *l1_ratio*입니다. 이때 alpha 값은 Ridge와 Lasso 클래스의 alpha 값과는 다릅니다. 

<br>

엘라스틱넷의 규제는 **a * L1 + b * L2** 로 정의됩니다. 

따라서,ㅤㅤ**alphaㅤㅤ--->ㅤㅤa+b**

ㅤㅤㅤㅤㅤ **l1_ratioㅤㅤ--->ㅤㅤa/(a+b)** 

<br>

펭귄 몸무게 예측 예제를 ElasticNet 클래스를 이용해 다시 예측해보고, 
alpha 값의 변화에 따른 RMSE와 엘라스틱넷 회귀 계수 값의 변화를 살펴보겠습니다.

In [123]:
# alpha 값의 변화에 따른 평균 RMSE의 변화

from sklearn.linear_model import ElasticNet
from sklearn.model_selection import cross_val_score

print('####### ElasticNet #######')

elastic_alphas = [0.07, 0.1, 0.5, 1, 3]

for alpha in elastic_alphas :
    elastic = ElasticNet(alpha = alpha, l1_ratio=0.7)

    neg_mse_scores = cross_val_score(elastic, feature, target, scoring="neg_mean_squared_error", cv = 5)
    avg_rmse = np.mean(np.sqrt(-1 * neg_mse_scores))
    print('alpha {0} 일 때 5 folds 의 평균 RMSE : {1:.3f} '.format(alpha, avg_rmse))

####### ElasticNet #######
alpha 0.07 일 때 5 folds 의 평균 RMSE : 364.375 
alpha 0.1 일 때 5 folds 의 평균 RMSE : 363.698 
alpha 0.5 일 때 5 folds 의 평균 RMSE : 365.521 
alpha 1 일 때 5 folds 의 평균 RMSE : 370.484 
alpha 3 일 때 5 folds 의 평균 RMSE : 381.549 


<br>

In [124]:
# alpha 값의 변화에 따른 회귀 계수의 변화

coeff_df = pd.DataFrame()
elastic_alphas = [0.07, 0.1, 0.5, 1, 3]

for pos, alpha in enumerate(elastic_alphas) :
    elastic = ElasticNet(alpha = alpha, l1_ratio=0.7)
    elastic.fit(feature, target)
 
    coeff = pd.Series(data=elastic.coef_, index=feature.columns)
    colname = 'alpha:'+str(alpha)
    coeff_df[colname] = coeff

sort_column = 'alpha:'+str(elastic_alphas[0])
coeff_df.sort_values(by=sort_column, ascending=False)

alpha:0.07   alpha:0.1   alpha:0.5     alpha:1    alpha:3
Flipper Length (mm)   32.285171   33.231496   38.730608   40.917988  43.206146
Clutch Completion     -1.962477   -1.018081   -0.097313   -0.490740  -0.000000
Culmen Length (mm)    -2.575066   -0.755154    8.147848   10.977077  13.519085
Delta 13 C (o/oo)     -6.167127   -5.644318   -2.238545   -0.000000   3.856317
Culmen Depth (mm)    -19.140550  -17.494857   -0.867532    5.864574   9.469335
Delta 15 N (o/oo)    -41.051653  -40.874741  -37.863638  -33.301323 -22.226774
Island              -107.734340 -110.150506 -103.458081  -83.928568 -44.580786
Species             -145.149907 -124.785793  -38.513740  -18.166085  -4.457717
Sex                 -477.821943 -447.156846 -248.048586 -162.650139 -69.726675

<Br>

alpha 값이 0.07, 0.1, 0.5, 1, 3으로 변화함에 따라 평균 RMSE의 값과 회귀 계수가 어떻게 변화하는지 살펴보았습니다.

alpha가 0.07일 때 334.074로 가장 좋은 평균 RMSE를 보여줍니다.

릿지 회귀와 라쏘 회귀의 평균 RMSE 값보다는 약간 떨어지는 수치지만, LinearRegression의 평균 RMSE인 336.434보다는 향상되었습니다.

이렇게 규제 선형 회귀의 가장 대표적인 기법인 릿지, 라쏘, 엘라스틱넷 회귀를 살펴보았습니다. 

이들 중 어떤 것이 가장 좋은지는 상황에 따라 다르기 때문에 각각의 알고리즘에서 하이퍼 파라미터를 변경해 가면서 최적의 예측 성능을 찾아내야 합니다.

<br>